File: clients_java/common/src/org/openamq/framing/ProtocolInitiation.java
package org.openamq.framing;

import org.apache.mina.common.ByteBuffer;
import org.apache.mina.common.IoSession;
import org.apache.mina.filter.codec.demux.MessageDecoderResult;
import org.apache.mina.filter.codec.demux.MessageDecoder;
import org.apache.mina.filter.codec.ProtocolDecoderOutput;
import org.openamq.AMQException;

public class ProtocolInitiation extends AMQDataBlock implements EncodableAMQDataBlock
{
    public char[] header = new char[]{'A','M','Q','P'};
    // TODO: generate these constants automatically from the xml protocol spec file
    public byte protocolClass = 1;
    public byte protocolInstance = 1;
    public byte protocolMajor = 10;
    public byte protocolMinor = 3;

    public long getSize()
    {
        return 4 + 1 + 1 + 1 + 1;
    }

    public void writePayload(ByteBuffer buffer)
    {
        for (int i = 0; i < header.length; i++)
        {
            buffer.put((byte) header[i]);
        }
        buffer.put(protocolClass);
        buffer.put(protocolInstance);
        buffer.put(protocolMajor);
        buffer.put(protocolMinor);
    }

    public void populateFromBuffer(ByteBuffer buffer) throws AMQException
    {
        throw new AMQException("Method not implemented");
    }

    public static class Decoder implements MessageDecoder
    {
        private boolean _disabled = false;

        public MessageDecoderResult decodable(IoSession session, ByteBuffer in)
        {
            if (_disabled)
            {
                return MessageDecoderResult.NOT_OK;
            }
            if (in.remaining() < 8)
            {
                return MessageDecoderResult.NEED_DATA;
            }
            else
            {
                char[] expected = new char[]{'A', 'M', 'Q', 'P'};
                for (int i = 0; i < 4; i++)
                {
                    if (((char) in.get()) != expected[i])
                    {
                        return MessageDecoderResult.NOT_OK;
                    }
                }
                return MessageDecoderResult.OK;
            }
        }

        public MessageDecoderResult decode(IoSession session, ByteBuffer in, ProtocolDecoderOutput out)
            throws Exception
        {
            byte[] header = new byte[4];
            in.get(header);
            ProtocolInitiation pi = new ProtocolInitiation();
            pi.header = new char[]{'A','M','Q','P'};
            byte temp = pi.protocolClass;
            pi.protocolClass = in.get();
            if (pi.protocolClass != temp)
            {
                throw new AMQProtocolVersionException("Protocol class " + temp + " was expected; received " +
                                                      pi.protocolClass);
            }
            temp = pi.protocolInstance;
            pi.protocolInstance = in.get();
            if (pi.protocolInstance != temp)
            {
                throw new AMQProtocolVersionException("Protocol instance " + temp + " was expected; received " +
                                                      pi.protocolInstance);
            }
            temp = pi.protocolMajor;
            pi.protocolMajor = in.get();
            if (pi.protocolMajor != temp)
            {
                throw new AMQProtocolVersionException("Protocol major version " + temp + " was expected; received " +
                                                      pi.protocolMajor);
            }
            temp = pi.protocolMinor;
            pi.protocolMinor = in.get();
            if (pi.protocolMinor != temp)
            {
                throw new AMQProtocolVersionException("Protocol minor version " + temp + " was expected; received " +
                                                      pi.protocolMinor);
            }
            out.write(pi);
            return MessageDecoderResult.OK;
        }

        public void setDisabled(boolean disabled)
        {
            _disabled = disabled;
        }
    }
}

File: dotnet/xmscommon/framing/ProtocolInitiation.cs
using System;
using jpmorgan.mina.common;
using jpmorgan.mina.filter.codec;
using jpmorgan.mina.filter.codec.demux;
using log4net;

namespace OpenAMQ.Framing
{
    public class ProtocolInitiation : IDataBlock, IEncodableAMQDataBlock
    {
        public char[] Header = new char[]{'A','M','Q','P'};
        public byte ProtocolClass = 1;
        public byte ProtocolInstance = 1;
        public byte ProtocolMajor = 10;
        public byte ProtocolMinor = 3;

        public uint Size
        {
            get
            {
                return 4 + 1 + 1 + 1 + 1;
            }
        }

        public void WritePayload(ByteBuffer buffer)
        {
            foreach (char c in Header)
            {
                buffer.Put((byte) c);
            }
            buffer.Put(ProtocolClass);
            buffer.Put(ProtocolInstance);
            buffer.Put(ProtocolMajor);
            buffer.Put(ProtocolMinor);
        }

        /// <summary>
        /// Populates from buffer.
        /// </summary>
        /// <param name="buffer">The buffer.</param>
        public void PopulateFromBuffer(ByteBuffer buffer)
        {
            throw new AMQException("Method not implemented");
        }

        public class Decoder : IMessageDecoder
        {
            private bool _disabled = false;

            public MessageDecoderResult Decodable(ISession session, ByteBuffer inbuf)
            {
                if (_disabled)
                {
                    return MessageDecoderResult.NOT_OK;
                }
                if (inbuf.Remaining < 8)
                {
                    return MessageDecoderResult.NEED_DATA;
                }
                else
                {
                    char[] expected = new char[]{'A', 'M', 'Q', 'P'};
                    for (int i = 0; i < 4; i++)
                    {
                        if (((char) inbuf.Get()) != expected[i])
                        {
                            return MessageDecoderResult.NOT_OK;
                        }
                    }
                    return MessageDecoderResult.OK;
                }
            }

            /// <summary>
            /// Decodes the specified session.
            /// </summary>
            /// <param name="session">The session.</param>
            /// <param name="inbuf">The inbuf.</param>
            /// <param name="outbuf">The outbuf.</param>
            /// <returns></returns>
            /// <exception cref="ProtocolViolationException">thrown if the frame violates the protocol</exception>
            public MessageDecoderResult Decode(ISession session, ByteBuffer inbuf, IProtocolDecoderOutput output)
            {
                byte[] header = new byte[4];
                inbuf.Get(header);
                ProtocolInitiation pi = new ProtocolInitiation();
                pi.Header = new char[]{'A','M','Q','P'};
                pi.ProtocolClass = inbuf.Get();
                pi.ProtocolInstance = inbuf.Get();
                pi.ProtocolMajor = inbuf.Get();
                pi.ProtocolMinor = inbuf.Get();
                output.Write(pi);
                return MessageDecoderResult.OK;
            }

            public bool Disabled
            {
                set
                {
                    _disabled = value;
                }
            }
        }
    }
}
